servicex_app/servicex_app/resources/internal/transformer_file_complete.py

```diff
@@ -71,57 +115,55 @@ def put(self, request_id):
         return "Ok"
 
     @staticmethod
-    @retry(stop=stop_after_attempt(3),
-           wait=wait_exponential_jitter(initial=0.1, max=30),
-           before_sleep=before_sleep_log(current_app.logger, logging.INFO),
-           after=after_log(current_app.logger, logging.INFO),
-           )
-    def record_file_complete(logger: Logger, request_id: str, info: dict[str, str]) -> TransformRequest | None:
-        transform_req = TransformRequest.lookup(request_id)
-        if transform_req is None:
-            msg = f"Request not found with id: '{request_id}'"
-            logger.error(msg, extra={'requestId': request_id})
-            return None
+    @file_complete_ops_retry
+    def record_file_complete(session: Session, logger: Logger, request_id: str,
+                             info: dict[str, str]) -> TransformRequest | None:
 
-        if info['status'] == 'success':
-            TransformRequest.file_transformed_successfully(request_id)
-        else:
-            TransformRequest.file_transformed_unsuccessfully(request_id)
+        with session.begin():
+            # Lock the row for update
+            transform_req = session.query(TransformRequest).filter_by(
+                request_id=request_id).with_for_update().one_or_none()
+
+            if transform_req is None:
+                msg = f"Request not found with id: '{request_id}'"
+                logger.error(msg, extra={'requestId': request_id})
+                return None
+
+            if info['status'] == 'success':
+                transform_req.files_completed += 1
+            else:
+                transform_req.files_failed += 1
+
+        session.flush()  # Flush the changes to the database
```

Comment:
What is the state of the DB if the transaction attempt fails three times? Do we again go out of sync?

What isolation level are we working at? (Are we certainly taking a row lock, or could the transaction fail because it's been updated somewhere else?)

Reply:
There are three operations. One of them is on table of transform results. Each file complete has it's own row, so there is no contention there.

The part that causes us the most concern is the number of files processed column in the TransformTable. That's the one I put the heaviest level of locking on. 

The other interesting one is declaring that the last file has been processed and so we can call the transform complete and shut down the transformers. I would argue that's not catastrophic if it happens twice (It is a big deal when it never happens!)

My biggest concern is deadlock, which is why I make the update of the TransformResult table a different transaction from the update of files remaining. So I think there is no chance for deadlock.

So, but good point about how to handle failures of each transaction/function call in this resource. Problem is, I don't think we have any good rollback options.

1. The update to the files_complete fails: We are in big trouble. There is no way to recover this and the transform will never quit.
2. The update to the transform_result fails... I don't think this will have much impact. As Illija points out I'm not sure who uses this table since we just scan the bucket for files to return in the client.
3. The status update fails. This is also a big deal, but not sure what we could do about it.

I suppose I should put some `try`/`except` blocks in this code to manage this, but as I say I don't know what to do if we catch an exception (and a function fails after three tries)